documentation/tuning_guide.markdown

```diff
@@ -0,0 +1,156 @@
+---
+layout: default
+title: "Puppet Server: Tuning Guide"
+canonical: "/puppetserver/latest/tuning_guide.html"
+---
+
+Puppet Server provides many configuration options that can be used to tune the
+server for maximum performance and hardware resource utilization.  In this guide,
+we'll highlight some of the most important settings that you can use to get
+the best performance in your environment.
+
+## Puppet Server and JRuby
+
+Before you begin tuning your configuration, it's helpful to have a little bit
+of context on how Puppet Server uses JRuby to handle incoming HTTP requests from
+your Puppet agents.
+
+When Puppet Server starts up, it creates a pool of JRuby interpreters to use
+as workers when we need to execute some of the Puppet Ruby code.  You can think
+of these almost as invidual Ruby "virtual machines" that are controlled by
+Puppet Server; it's not entirely dissimilar to the way that Passenger spawns
+several MRI Ruby processes to hand off work to.
+
+Puppet Server isolates these JRuby instances so that they will only be allowed
+to handle one request at a time.  This ensures that we do not encounter any
+concurrency issues, since the Ruby code is not thread-safe.  When an HTTP request
+comes in to Puppet Server and it determines that some Ruby code will need to be
+executed in order to handle the request, Puppet Server "borrows" a JRuby instance
+from the pool, uses it to do the work, and then "returns" it to the pool.  If
+there are no JRuby instances available in the pool at the time when a request
+comes in (presumably because all of the JRuby instances are already in use handling
+other requests), Puppet Server will block the request until one becomes available.
+
+(In the future, this approach will allow us to do some really powerful things
+such as creating multiple pools of JRubies and isolating each of your Puppet
+environments to a single pool, to ensure that there is no pollution from one
+Puppet environment to the next.)
+
+This brings us to the two most important settings that you can use to tune your
+Puppet Server.
+
+### Number of JRubies
+
+The most important setting that you can use to improve the throughput of your
+Puppet Server installation is the [`max-active-instances`](./configuration.html#puppetserver_conf)
```

Comment:
this link looks broken to me, but perhaps that's just an artifact of how ours docs are laid-out?